### ISSUE TITLE
Some notifications read improperly

### ISSUE THREAD

**Issue by cherryleafroad** [Nevolution/decorator-wechat#43]
I noticed that when somebody quotes a message that was using new wechat quote feature, a lot of messages I'm the notifications appear as `[Message]` instead of the actual content. I turned off nevo to see if it's wechat fault, but wechat displayed notification just fine when quoting messages. This doesn't seem to happen to regular unquoted messages

I think it with happened when somebody sent a link too. 

Using wechat 8.0.1, nevo 6.2, Android 11, pixel 2 xl

**Comment by cherryleafroad** [comment]
I tried to fix this by myself, but it turns out it's not that easy. It seems to be a WeChat bug.

What I learned is that in `WeChatDecorator.java` @ here
```conversation.ext = IGNORE_CAR_EXTENDER ? null : new Notification.CarExtender(n).getUnreadConversation();```

It is already being interpreted as `[Message]` even at this point. It appears to me that this is WeChat's fault for not providing proper data for the car bundle, because when `CarExtender` reaches the end
`mUnreadConversation = UnreadConversation.getUnreadConversationFromBundle(b);`, `getUnreadConversationFromBundle()` will return `[Message]` for all affected entries. After debugging the inside of `getUnreadConversationFromBundle()`, I can confirm the data was already affected.

The ticker remains unaffected.

**Comment by cherryleafroad** [comment]
I patched the problem by creating a Conversation History class and caching the conversation. If it sees a "[Message]", it will grab the data from the ticker, car bundle (if it's not corrupted), its internal cache (already calculated), or if all else fails calculate it from notification history (using `getArchivedNotifications()`; unfortunately I have no choice but to rely on this, as there's no other way to get the back data, e.g. in case the application didn't have it cached before and car bundle is corrupted). Because I'm not sure how good my solution is - and I'm sure you can come up with a better one, so I didn't send a pull request for it. However, I'll link to my solution for you to either use or help to derive your own from. [Find it here](https://github.com/cherryleafroad/decorator-wechat/tree/message-patch)

**Comment by oasisfeng** [comment]
Ye, Conversation History is also the plan in my mind, to improve the conversation experience and fix a lot more issues.

Thanks for sharing your thoughts on this.

**Comment by cherryleafroad** [comment]
Certainly!! Conversation History is really helpful. I even used it to implement recalled messages functionality (you can find in my dev branch)

**Comment by oasisfeng** [comment]
I'll review your contribution as soon as I had time.

**Comment by Marukon** [comment]
> I patched the problem by creating a Conversation History class and caching the conversation. If it sees a "[Message]", it will grab the data from the ticker, car bundle (if it's not corrupted), its internal cache (already calculated), or if all else fails calculate it from notification history (using `getArchivedNotifications()`; unfortunately I have no choice but to rely on this, as there's no other way to get the back data, e.g. in case the application didn't have it cached before and car bundle is corrupted). Because I'm not sure how good my solution is - and I'm sure you can come up with a better one, so I didn't send a pull request for it. However, I'll link to my solution for you to either use or help to derive your own from. [Find it here](https://github.com/cherryleafroad/decorator-wechat/tree/message-patch)

Could you release a apk, please?

**Comment by cherryleafroad** [comment]
> Could you release a apk, please?
@Marukon Due to the way the program is implemented, the apk wouldn't work for you due to `sharedUserId` and signing keys. As such, the apk would be incompatible with your Nevo, unless you decide to re-sign your Nevo and change the sharedUserId + processes in the manifest. Your best bet is to wait for the author to check / merge / implement my solution (or a variation thereof)

TLDR: Possible if you want to do a lot of work to make it compatible, otherwise, not possible
